securedrop/journalist_app/api2/__init__.py

```
@@ -14,18 +15,57 @@
 PREFIX_MAX_LEN = inspect(Source).columns["uuid"].type.length
 
 
-def all_sources() -> Query:
-    """
-    Return a base query for all ``Sources`` with eager loading of their metadata
-    and collections.
-    """
+# Used by `{Submission,Reply}_query_options()` when called directly rather than
+# via `Source_query_options()`.
+LOADER_BASE = {
+    Submission: Load(Submission),
+    Reply: Load(Reply),
+}
+
+
+def Submission_query_options(base: Load = LOADER_BASE[Submission]) -> Tuple:
+    configure_mappers()
+    return (
+        base.joinedload(Submission.source),  # type: ignore[attr-defined]
+        base.joinedload(Submission.seen_files).joinedload(SeenFile.journalist),  # type: ignore[attr-defined]
+        base.joinedload(Submission.seen_messages).joinedload(SeenMessage.journalist),  # type: ignore[attr-defined]
+    )
+
+
+def Reply_query_options(base: Load = LOADER_BASE[Reply]) -> Tuple:
+    configure_mappers()
     return (
-        Source.query.options(joinedload(Source.star))
-        .options(joinedload(Source.submissions))
-        .options(joinedload(Source.replies))
+        base.joinedload(Reply.source),  # type: ignore[attr-defined]
+        base.joinedload(Reply.journalist),  # type: ignore[attr-defined]
+        base.joinedload(Reply.seen_replies).joinedload(SeenReply.journalist),  # type: ignore[attr-defined]
     )
 
 
+def Source_query_options() -> Tuple:
+    configure_mappers()
+    return (
+        joinedload(Source.star),
+        *Submission_query_options(joinedload(Source.submissions)),
+        *Reply_query_options(joinedload(Source.replies)),
+    )
+
+
+EAGER_BUNDLES = {
+    Source: lambda: Source_query_options(),
+    Submission: lambda: Submission_query_options(),
+    Reply: lambda: Reply_query_options(),
+}
+
+
+def eager_query(model: db.Model) -> Query:
+    """
+    Return ``model.query`` with the appropriate eager-loading options applied.
+    Falls back to plain ``model.query`` if no bundle is registered.
```

I don't think we should have a silent fallback; if you use `eager_query()` it should actually do something or loudly complain (i.e. an exception) that it's not going to make a difference so that you know to configure it for that type. Or maybe instead of an exception we can codify it at the type-level?

---

Done, with the caveats noted, in fb0b3e003b5e250107b730916cf180e54aac4c5c—and proven useful by 44419da1060c86a8934231528ffcce9137f8d1bc.  :-)